common/src/main/java/com/linecorp/centraldogma/common/PathPatternBuilder.java

```diff
@@ -0,0 +1,125 @@
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Builds a new {@link PathPattern}.
+ *
+ * <h2>Example</h2>
+ * <pre>{@code
+ * final PathPattern pathPattern =
+ *         PathPattern.builder()
+ *                    .startsWith("/foo/bar")
+ *                    .contains("/ext")
+ *                    .extension("json")
+ *                    .build();
+ * }</pre>
+ */
+public final class PathPatternBuilder {
+    @Nullable
+    private PathPatternOption startPattern;
+    private final List<PathPatternOption> innerPatterns = new ArrayList<>();
+    @Nullable
+    private PathPatternOption endPattern;
+
+    PathPatternBuilder() {}
+
+    /**
+     * Adds {@link PathPatternOptions#ENDS_WITH}.
+     */
+    public PathPatternBuilder endsWith(String filename) {
+        endPattern = PathPatternOptions.ENDS_WITH.apply(filename);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#EXTENSION}.
+     */
+    public PathPatternBuilder extension(String extension) {
+        endPattern = PathPatternOptions.EXTENSION.apply(extension);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#STARTS_WITH}.
+     */
+    public PathPatternBuilder startsWith(String dirPath) {
+        startPattern = PathPatternOptions.STARTS_WITH.apply(dirPath);
+        return this;
+    }
+
+    /**
+     * Adds {@link PathPatternOptions#CONTAINS}.
+     */
+    public PathPatternBuilder contains(String dirPath) {
+        innerPatterns.add(PathPatternOptions.CONTAINS.apply(dirPath));
+        return this;
+    }
+
+    /**
+     * Compose one pathPattern from a list of {@code patterns}.
+     */
+    private static String combine(List<PathPattern> patterns) {
+        final StringBuilder sb = new StringBuilder();
+        for (final Iterator<PathPattern> i = patterns.iterator(); i.hasNext();) {
+            if (sb.length() == 0) {
+                // left should end with "/**"
+                sb.append(i.next().patternString());
+            } else {
+                // right should start with "/**/"
+                sb.append(i.next().patternString().substring(3));
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Returns a newly-created {@link PathPattern} based on the options of this builder.
+     */
+    public PathPattern build() {
+        final ImmutableList.Builder<PathPatternOption> optionsBuilder = ImmutableList.builder();
+        if (startPattern != null) {
+            optionsBuilder.add(startPattern);
+        }
+        optionsBuilder.addAll(innerPatterns);
+        if (endPattern != null) {
+            optionsBuilder.add(endPattern);
+        }
+        final ImmutableList<PathPatternOption> options = optionsBuilder.build();
+
+        checkArgument(!options.isEmpty(), "Requires at least one pattern to build in PathPatternBuilder");
```

Comment:
If a method validates a value that the method directly takes, `IllegalArgumentException` would be fine. `options` is created from the values given before. In that case, we normally use `IllegalStateException`.
```suggestion
        checkState(!options.isEmpty(), "Requires at least one pattern to build in PathPatternBuilder");
```